### PR TITLE
ARD page creation

### PR DESCRIPTION
**Pull request by amoriartyCH**
-Created Accounts Reference Page
-Changed Next&Prev controllers to fit page in correctly
-Added unit tests

### CODE REVIEW

File: src/main/java/uk/gov/companieshouse/web/accounts/controller/smallfull/AccountsReferenceDateController.java [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+import java.util.List;
+
+@Controller
+@NextController(DirectorsReportQuestionController.class)
+@PreviousController(AccountsReferenceDateQuestionController.class)
+@RequestMapping("/company/{companyNumber}/transaction/{transactionId}/company-accounts/{companyAccountsId}/small-full/accounts-reference-date")
+public class AccountsReferenceDateController extends BaseController implements ConditionalController {
+
+    @Autowired
+    NavigatorService navigatorService;
```
Comment (BSaunders95) [inline src/main/java/uk/gov/companieshouse/web/accounts/controller/smallfull/AccountsReferenceDateController.java:43]: These could be private

File: src/main/java/uk/gov/companieshouse/web/accounts/controller/smallfull/AccountsReferenceDateController.java [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+            AccountsReferenceDate accountsReferenceDate = new AccountsReferenceDate();
+
+            ApiClient apiClient = apiClientService.getApiClient();
+            CompanyProfileApi companyProfile = companyService.getCompanyProfile(companyNumber);
+            SmallFullApi smallFullAccounts = smallFullService.getSmallFullAccounts(apiClient, transactionId, companyAccountsId);
+
+            accountsReferenceDate.setPastDates(getPastValidDates(companyProfile.getAccounts().getNextAccounts().getPeriodEndOn()));
+            accountsReferenceDate.setFutureDates(getFutureValidDates(companyProfile.getAccounts().getNextAccounts().getPeriodEndOn()));
+
+            if(!companyProfile.getAccounts().getNextAccounts().getPeriodEndOn().equals(smallFullAccounts.getNextAccounts().getPeriodEndOn())) {
```
Comment (BSaunders95) [inline src/main/java/uk/gov/companieshouse/web/accounts/controller/smallfull/AccountsReferenceDateController.java:79]: Could we maybe abstract this into a private method? No fault of yours, it's just not the easiest to read!

File: src/main/java/uk/gov/companieshouse/web/accounts/controller/smallfull/AccountsReferenceDateController.java [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+    protected String getTemplateName() {
+        return "smallfull/accountsReferenceDate";
+    }
+
+    @Override
+    public boolean willRender(String companyNumber, String transactionId, String companyAccountsId)
+            throws ServiceException {
+
+        CompanyAccountsDataState companyAccountsDataState = getStateFromRequest(request);
+        return BooleanUtils.isTrue(!companyAccountsDataState.getHasConfirmedAccountingReferenceDate());
```
Comment (BSaunders95) [inline src/main/java/uk/gov/companieshouse/web/accounts/controller/smallfull/AccountsReferenceDateController.java:127]: Could we flip this and use `BooleanUtils#isFalse` instead?

File: src/main/java/uk/gov/companieshouse/web/accounts/controller/smallfull/AccountsReferenceDateController.java [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+    }
+
+    private List<LocalDate> getFutureValidDates(LocalDate periodEndOn) {
+
+        List<LocalDate> futureValidDates = new ArrayList<>();
+
+
+        futureValidDates.add(periodEndOn);
```
Comment (BSaunders95) [inline src/main/java/uk/gov/companieshouse/web/accounts/controller/smallfull/AccountsReferenceDateController.java:135]: Careful here - we don't want period end as an option as the user has said they want to change their period end

File: src/main/java/uk/gov/companieshouse/web/accounts/controller/smallfull/AccountsReferenceDateController.java [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+
+    @Override
+    public boolean willRender(String companyNumber, String transactionId, String companyAccountsId)
+            throws ServiceException {
+
+        CompanyAccountsDataState companyAccountsDataState = getStateFromRequest(request);
+        return BooleanUtils.isTrue(!companyAccountsDataState.getHasConfirmedAccountingReferenceDate());
+    }
+
+    private List<LocalDate> getFutureValidDates(LocalDate periodEndOn) {
```
Comment (BSaunders95) [inline src/main/java/uk/gov/companieshouse/web/accounts/controller/smallfull/AccountsReferenceDateController.java:130]: Could we maybe abstract this logic into, say, company service? The logic split would make unit testing simpler / more thorough and I think the controller may be doing too much at present - especially when the logic to remove future dates is added.

File: src/main/java/uk/gov/companieshouse/web/accounts/model/smallfull/AccountsReferenceDate.java [hunk 6]
```diff
@@ -0,0 +1,39 @@
+package uk.gov.companieshouse.web.accounts.model.smallfull;
+
+import javax.validation.constraints.NotNull;
+import java.time.LocalDate;
+import java.util.List;
+
+public class AccountsReferenceDate {
+
+    @NotNull(message = "{accountsReferenceDate.selectionNotMade}")
+    LocalDate chosenDate;
```
Comment (BSaunders95) [inline src/main/java/uk/gov/companieshouse/web/accounts/model/smallfull/AccountsReferenceDate.java:10]: Access modifiers (sorry!)

File: src/main/resources/templates/smallfull/accountsReferenceDate.html [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+
+                <div class="govuk-form-group"
+                     th:classappend="${#fields.hasErrors('chosenDate')} ? 'govuk-form-group--error' : ''">
+                    <div class="govuk-radios">
+                        <div class="govuk-grid-row">
+                            <div class="govuk-grid-column-one-half">
+                                <th:block th:each="pastDate, stat : ${accountsReferenceDate.pastDates}">
+                                    <div class="govuk-radios__item">
+                                        <input class="govuk-radios__input" type="radio"
+                                               th:id="${stat.index}"
```
Comment (BSaunders95) [inline src/main/resources/templates/smallfull/accountsReferenceDate.html:36]: This id is going to match the one for future dates (i.e. there will be 2 radios on the page with `id="0"`)

File: src/main/java/uk/gov/companieshouse/web/accounts/controller/smallfull/AccountsReferenceDateController.java (hunk 1 above)
Comment (amoriartyCH) [reply src/main/java/uk/gov/companieshouse/web/accounts/controller/smallfull/AccountsReferenceDateController.java]: Fixed

File: src/main/java/uk/gov/companieshouse/web/accounts/controller/smallfull/AccountsReferenceDateController.java (hunk 3 above)
Comment (amoriartyCH) [reply src/main/java/uk/gov/companieshouse/web/accounts/controller/smallfull/AccountsReferenceDateController.java:127]: Flipped as suggested

File: src/main/java/uk/gov/companieshouse/web/accounts/controller/smallfull/AccountsReferenceDateController.java (hunk 4 above)
Comment (amoriartyCH) [reply src/main/java/uk/gov/companieshouse/web/accounts/controller/smallfull/AccountsReferenceDateController.java:135]: Removed periodEndOn from list and added a 7th day to display (as prototype does).

File: src/main/java/uk/gov/companieshouse/web/accounts/model/smallfull/AccountsReferenceDate.java (hunk 6 above)
Comment (amoriartyCH) [reply src/main/java/uk/gov/companieshouse/web/accounts/model/smallfull/AccountsReferenceDate.java:10]: Fixed

File: src/main/java/uk/gov/companieshouse/web/accounts/controller/smallfull/AccountsReferenceDateController.java (hunk 2 above)
Comment (amoriartyCH) [reply src/main/java/uk/gov/companieshouse/web/accounts/controller/smallfull/AccountsReferenceDateController.java:78]: Implemented

File: src/main/java/uk/gov/companieshouse/web/accounts/controller/smallfull/AccountsReferenceDateController.java [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+    private SmallFullServiceImpl smallFullService;
+
+    @Autowired
+    private HttpServletRequest request;
+
+    private static final String ACCOUNTS_REFERENCE_DATE = "accountsReferenceDate";
+
+
+    @GetMapping
+    public String getAccountsReferenceDateQuestion(@PathVariable String companyNumber,
```
Comment (mreis-ch) [inline src/main/java/uk/gov/companieshouse/web/accounts/controller/smallfull/AccountsReferenceDateController.java:60]: Should the Question bit be removed?

**Pull request merged**